src/imgui/Shortcuts.cc

```diff
@@ -0,0 +1,136 @@
+#include "Shortcuts.hh"
+#include "imgui_internal.h"
+
+#include "SettingsConfig.hh"
+
+#include <array>
+
+namespace openmsx {
+
+//-- (Only) the table in this section will be edited when adding new shortcuts
+
+struct AllShortcutInfo {
+    ShortcutIndex index;
+    ImGuiKeyChord keyChord;
+    bool local;
+    bool repeat;
+    std::string_view name; // used in settings.xml
+    std::string_view description; // shown in GUI
+};
+
+// When adding a new Shortcut, we only:
+//  * Add a new value in the above 'enum ShortcutIndex'
+//  * Add a single line in this table
+// Note: if you look at the generated code on the right, this table does NOT end up in the generated code
+//       instead it's only used to calculate various other tables (that do end up in the generated code)
+//       those 'calculations' are all done at compile-time
+// Note: the first column is ONLY used to verify (at-compile-time) whether the rows in this
+//       table are in the same order as the values in ShortcutIndex
+static constexpr auto allShortcutInfo = std::to_array<AllShortcutInfo>({
+    {GOTO_MEMORY_ADDRESS, ImGuiMod_Ctrl | ImGuiKey_G,                  false,  false, "goto_memory_address",       "Go to address in hex viewer"},
+    {STEP,                ImGuiKey_F6,                                 false,  true,  "step",                      "Single step in debugger"},
+    {BREAK,               ImGuiKey_F7,                                 false,  false, "break",                     "Break CPU emulation in debugger"},
+    {GOTO_DISASM_ADDRESS, ImGuiMod_Ctrl | ImGuiMod_Shift | ImGuiKey_G, false,  false, "goto_disasm_address",       "Go to address in dissassembler"},
+    {MOVE_UP_MEMORY,      ImGuiKey_UpArrow,                            true,   true,  "move_up_memory_address",    "Move up in hex viewer"},
+    {MOVE_DOWN_MEMORY,    ImGuiKey_DownArrow,                          true,   true,  "move_down_memory_address",  "Move down in hex viewer"},
+    {MOVE_LEFT_MEMORY,    ImGuiKey_LeftArrow,                          true,   true,  "move_left_memory_address",  "Move left in hex viewer"},
+    {MOVE_RIGHT_MEMORY,   ImGuiKey_RightArrow,                         true,   true,  "move_right_memory_address", "Move right hex viewer"},
```

Comment:
Detail:
I understand these are only toy-examples (not meant for the final version). But here's a remark anyway:
* I think `CTRL+G` must be a local binding, because there can be multiple hex-editors open at the same time, and you don't want them all to react to `CTRL+G`.
* And then it becomes possible to reuse `CTRL+G` also for `GOTO_DISASM_ADDRESS`.

Reply:
> I think `CTRL+G` must be a local binding, because there can be multiple hex-editors open at the same time, and you don't want them all to react to `CTRL+G`.

I think the shortcut widget should be dynamically populated by the existing widgets on the screen. If we create a new hex editor, it should appear in the shortcut widget (with an initial binding set to None).

Reply:
That's possible. But personally I don't think it's a good idea.

For example I open the memory viewer (and possibly setup my preferred shortcuts for it). Then I open a 2nd memory editor (so a 2nd view on the same debuggable). Then it would be weird if I have to redefine my shortcuts (even worse: I have to pick new unique keys for each). It would also be weird if the default shortcuts only work in the 1st editor I open but not in the 2nd. And actually the same story for 2 editors on different debuggables. It's the same widget, then why doesn't it react the same for all?

I also don't know of any applications that behave like this (different bindings for different instances of the same widget). For example I can open multiple terminals in my system. I did configure some custom shortcuts for my terminal. Personally I would find it very annoying if those shortcuts don't work the same in all my terminals. And obviously the shortcuts only have effect in the single terminal that has focus.

In general (I mean also outside of openMSX) global shortcuts are not that common, but they do exist, e.g. a global shortcut to open a new terminal. But global shortcuts that interact with specific instances of widgets/applications are very uncommon. I think it will be the same in openMSX. In fact the only useful global shortcut I can think of right now are the various "step*" debugger functions. (And maybe those could even be implemented as hotkeys, not as ImGui shortcuts. But let's keep global GUI shortcuts: for some users, in some specific case, they might be useful, I just don't think we should offer many by default).

It would also needlessly complicate the design, e.g. then it's no longer possible to have a static table with all possible shortcuts. And e.g. during loading of `settings.xml` you have to keep track of shortcuts that don't exist yet (e.g. because the corresponding hex-editor has not been created yet).

Reply:
The default shortcut still works on all instances of the same widget, but they won't receive it while the original widget is open. And maybe a very committed user that built himself a very customised debugging session would appreciate the flexibility to do this. But maybe we can think about that later.